src/Storages/MergeTree/ReplicatedMergeTreeMergeStrategyPicker.h

```diff
@@ -0,0 +1,66 @@
+#pragma once
+
+#include <common/types.h>
+// #include <Common/ZooKeeper/ZooKeeper.h>
+// #include <common/logger_useful.h>
+#include <optional>
+#include <mutex>
+#include <vector>
+#include <atomic>
+#include <boost/noncopyable.hpp>
+
+namespace DB
+{
+
+class StorageReplicatedMergeTree;
+struct ReplicatedMergeTreeLogEntryData;
+
+/// In some use cases merging can be more expensive than fetching
+/// (so instead of doing exactly the same merge cluster-wise you can do merge once and fetch ready part)
+/// Fetches may be desirable for other operational reasons (backup replica without lot of CPU resources).
+///
+/// That class allow to take a decisions about preferred strategy for a concreate merge.
+///
+/// Since that code is used in shouldExecuteLogEntry we need to be able to:
+/// 1) make decision fast
+/// 2) avoid excessive zookeeper operations
+///
+/// Because of that we need to cache some important things,
```

Comment:
Some notion that this class doesn't provide any strict guarantees. We may execute some merges when we have to wait or postpone some of them when we have to execute.